### PR TITLE
Added url root navigation

### PR DESCRIPTION
**Pull request by ghost**
For some reason the shortcut doesn't work but the command is registered and works. Not sure why that is happening.

### CODE REVIEW

**Comment by antoyo** [comment]
Thanks for your contribution.

The config file should be copied to `~/.config/titanium/config`.
Have you copied your changes in this file?

**Comment by ghost** [comment]
ah okay, gotcha, still figuring out the codebase. yeah it works

**Comment by antoyo** [comment]
Is going to the parent directory a feature you actually need?
I have not thought of this feature and I can some some uses case, so I'm okay with adding it.

For the default shortcut of `go-root-dir`, I think I'd use `gr`.
What do you think?

**Comment by ghost** [comment]
yeah there's a few more I'd like to throw in, for example [[ and ]] to detect the next and previous page like on blogs

chromium vim does this stuff and the gU bind is from there, but gr works too :)
https://github.com/1995eaton/chromium-vim

parent dir is nice for things like entering comment sections of reddit and then going back up to the subreddit, rather than to the hostname. Which doesn't work on reddit because it autoredirects, so that's why nth directory is nice to have as well as the ability to make site specific binds..

something like:
```
*://reddit.com/r/*/comments/* {
    map gu :go-parent-dir 2
}
```

**Comment by antoyo** [comment]
Let's keep `gU`: vimperator also uses that and I try to follow it.

By the way, you might want to experiment with `#[count]` parameter, even thought I don't use that in `titanium` yet.
[Here's an example.](https://github.com/antoyo/mg/blob/master/examples/main.rs#L248-L249)
So you might want to use that for going to the nth parent.
That will be useful to add in some places like scrolling, so if you want to contribute, that will be very much appreciated.

Also, similar to `[[` is `<C-a>` (with the opposite being `<C-x>`) to increment a number in the URL (useful for page count).

By the way, if you want me to create issues on GitHub so that you don't have to look at all the `TODO`s and `FIXME`s, please let me know.

Thanks again for your work.

**Comment by ghost** [comment]
alright, did a few more

File: config/keys [hunk 1]
```diff
@@ -11,6 +11,10 @@ nmap f follow
 nmap F win-follow
 nmap gg scroll-top
 nmap gi focus-input
+nmap gu go-parent-dir
+nmap gU go-root-dir
+nmap g- url-decrement
```
Comment (antoyo) [inline config/keys:16]: These shorcuts are `<C-x>` and `<C-a>` in vimperator.
Are they different in Chromium vim?

**Comment by antoyo** [comment]
Thanks for your great work.
It seems you `cherry-pick`ed instead of `rebase`d, the next time, please `rebase` with master: the diff shown here won't show changes unrelevant to you pull request.

**Comment by ghost** [comment]
apologies about the cherry-pick.

and yeah, to me some of cvim shortcuts make more sense and are familiar, I'm actually not even sure where to find the defaults for vimperator. 

I think ideally there should be two default key configs: `keys-vimperator` and `keys-cvim`. Then you can just have the default inside `config` be `include keys-vimperator`. On the faq / homepage you could have a cheatsheet comparing the two (or three, if you want to rock a custom default).

**Comment by antoyo** [comment]
No problem.
I'm also not sure where to find vimperator shortcuts.
I think I looked at its source code to find them ^^ .
I'll test your changes tonight.
I'll also think about providing multiple keys configuration files.

**Comment by ghost** [comment]
okay, I also did some file organizing that I think will help a lot. do you want me to push it now or submit a new pull request?
```
├── app
│  ├── bookmarks.rs
│  ├── browser.rs
│  ├── clipboard.rs
│  ├── config.rs
│  ├── copy_paste.rs
│  ├── dialog.rs
│  ├── download.rs
│  ├── file_chooser.rs
│  ├── hints.rs
│  ├── mod.rs
│  ├── pass_filler.rs
│  ├── paths.rs
│  ├── popup.rs
│  ├── search_engine.rs
│  ├── server.rs
│  ├── test_utils.rs
│  └── url.rs
├── commands.rs
├── completers
│  ├── bookmark.rs
│  ├── file.rs
│  └── mod.rs
├── errors.rs
├── file.rs
├── main.rs
├── managers
│  ├── bookmarks.rs
│  ├── config_dir.rs
│  ├── mod.rs
│  ├── password.rs
│  └── popup.rs
├── message_server.rs
├── settings.rs
├── stylesheet.rs
├── urls.rs
└── views
   ├── download.rs
   ├── download_list.rs
   ├── mod.rs
   └── webview.rs
```

**Comment by antoyo** [comment]
I'll prefer that as another pull request since this one already contains a lot.
That looks good at first glance.
Where will be the `App` view?

**Comment by ghost** [comment]
I wasn't sure where to put it, so the mod.rs is pretty much the same at the moment. I'll probably take another look at `app/mod.rs` and attempt to migrate the view to `/views/app.rs`. I think `handle_command` definitely need to be at the top of the mod.rs, or the only thing in it, and then separate the view model from the event logic but I'm unsure how difficult or elegant it will be.

**Comment by antoyo** [comment]
I haven't looked at it yet, but I'll do it soon.
Is it ready to be merged?

By the way, the error we see at startup is because the web browser asks for the scroll percentage before the connection with the web extension was made. When the next version of the `glib` crate will be released, I'll use an [event listener](https://github.com/antoyo/webkit2gtk-webextension-rs/blob/master/src/auto/dom_event_target.rs#L23) over the scroll event, but that requires a [closure](https://github.com/gtk-rs/glib/blob/master/src/closure.rs) which was added recently into the `glib` crate.

Following my fix of yesterday (for the text sometimes inserted backwards in the status bar entry — which has also seemed to fix the slight slow down of the browser), I found there's a bug causing the download input dialog to not be shown: I'll try to fix that as soon as possible.
Since that was a somewhat involved change, other issues might had appear.

**Comment by ghost** [comment]
It should be yeah.

I've been looking at that duckduckgo menu bug and it's a head scratcher. That closure for scrolling should make it easy to implement smoothscroll so that's nice. Also, were you planning on implementing tabs and tab groups soon?

**Comment by antoyo** [comment]
Ok, I'll try your changes now.

For the hint positions in duckduckgo menu, it will be easy to fix since a [new feature was added in webkit2gtk](https://bugs.webkit.org/show_bug.cgi?id=163892).
I don't know in which version it will be added, thought.

There's already a [webkit setting for smooth scrolling](https://github.com/antoyo/titanium/blob/master/src/settings.rs#L76), but I've never tried it.

As for tabs, I don't need them since I have them in my window manager.
If you want them, please open an issue in [mg](https://github.com/antoyo/mg), which is the crate I made for the UI and is used by titanium, so that we can discuss on their design (mg is a library similar to [girara](https://pwmt.org/projects/girara/)). I created a library from scratch instead of wrapper `girara` because it is too unsafe to use.
An alternative to get tabs would be to used  a software like `tabbed` to get tabs.

**Comment by antoyo** [comment]
The `go-parent-dir` does not work well when the URL does not end with a `/`, for instance: https://github.com/antoyo/titanium/tree/master/config/stylesheets
It will go to https://github.com/antoyo/titanium/tree/master/ instead of https://github.com/antoyo/titanium/tree/master/config/.

**Comment by antoyo** [comment]
Also, the `url-decrement` and `url-increment` does not work for query parameters like `?page=1`.

**Comment by antoyo** [comment]
Thanks, I'll test that.

The code for URL update looks needlessly complicated, but I don't see a way to simplify it with the `url` crate: for instance, this [type](https://docs.rs/url/newest/url/struct.PathSegmentsMut.html) does not allow us to change a particular path segment.
And the `urlparse` crate would not help us, as far as I know.
I might write a url parsing crate that will allow us to simplify this code later.

File: src/app/mod.rs [hunk 2]
```diff
@@ -475,6 +497,8 @@ impl App {
             BookmarkDel => self.delete_bookmark(),
             BookmarkEditTags => self.edit_bookmark_tags(),
             ClearCache => self.clear_cache(),
+            ClickNextPage => self.server_send(InnerMessage::ClickNextPage()),
```
Comment (antoyo) [inline src/app/mod.rs:500]: I used to put the calls to `server_send()` in `src/app/server.rs` but that might not be needed anymore since most of them are one-liner.
Which way do you think is best?
I see two avantages to stick with the old way:

 * We don't need to fully qualify the message in case it's the same as a command.
 * This let this `handle_command()` method cleaner.

File: src/app/url.rs [hunk 3]
```diff
@@ -42,4 +44,55 @@ impl App {
...
         }
     }
+
+    /// Go up one directory in url
+    pub fn go_parent_directory(&self) {
+        let mut parent = String::new();
+
+        if let Some(ref url) = self.webview.widget().get_uri() {
+            if let Some(filename) = get_filename(url) {
+                if filename.is_empty() {
```
Comment (antoyo) [inline src/app/url.rs]: Do you think it is possible to simplify this code by using [`Url::path_segments_mut()`](https://docs.rs/url/newest/url/struct.Url.html#method.path_segments_mut) and [`PathSegmentsMut::pop()`](https://docs.rs/url/newest/url/struct.PathSegmentsMut.html#method.pop) (that might also requires [`PathSegmentsMut::pop_if_empty()`](https://docs.rs/url/newest/url/struct.PathSegmentsMut.html#method.pop_if_empty))?
I'm not sure how query parameters are handled: perhaps we need to also clear them.

File: src/app/url.rs [hunk 4]
```diff
@@ -42,4 +44,55 @@ impl App {
...
+        }
+    }
+
+    /// Go to the root directory or url hostname 
+    pub fn go_root_directory(&self) {
+        if let Some(ref url) = self.webview.widget().get_uri() {
+            if let Ok(base_url) = Url::parse(url) {
+                let root = &base_url[..Position::BeforePath];
+
+                // #TODO: Test on file:/// and edge cases
```
Comment (antoyo) [inline src/app/url.rs]: This works fine on `file:///`.
What edge cases did you have in mind?

File: src/urls.rs [hunk 5]
```diff
@@ -69,3 +69,61 @@ pub fn is_url(input: &str) -> bool {
...
+        if let Some(query) = url.query(){
+            let pairs = url.query_pairs().into_owned();
+
+            // ParseIntoOwned lacks DoubleEndedIterator for rev(), must be parsed left to right
+            let next = pairs.map(|(lhs, rhs)| {
+                if updated {
+                    lhs + "=" + &rhs
+                } else if let Ok(number) = rhs.parse::<i32>() {
+                    updated = true;
+                    lhs + "=" + (number + inc_offset).to_string().as_str()
```
Comment (antoyo) [inline src/urls.rs]: This is going to increment every number in the URL query string, right?
That might not be what we want.

File: titanium-web-extension/src/executor/mod.rs [hunk 6]
```diff
@@ -227,6 +231,41 @@ impl Executor {
...
+    fn click_prev_page(&mut self) {
+        let document = self.model.page.get_dom_document();
+        let regex = Regex::new(r"(?i:prev(ious)|back|newer|less|«|‹)|(?:<.+>)<(?:<.+>)").unwrap();
+
+        if let Some(document) = document {
+            if let Some(link) = match_pattern(&document, "a", regex) {
+                if let Ok(element) = link.clone().downcast::<DOMHTMLElement>() {
+                    self.click(element);
+                } else {
+                    warn!("click_next_page, Failed downcast on element");
```
Comment (antoyo) [inline titanium-web-extension/src/executor/mod.rs]: The [`wtry!` macro](https://github.com/antoyo/titanium/blob/master/titanium-web-extension/src/lib.rs#L94) already support showing an error similar to this code.

**Comment by ghost** [comment]
I agree with the url crate... it was an embarrassing function to have to commit. I took a look at the `get_base_url` function previously to improve it since it seemed a little over-engineered but there's not really a good way to use the library in an even slightly complex manner. It has a few ways to do the same thing but none are very friendly and do the thing I want it to do. I too see room for a better library. Maybe one with better granularity, but also have macros for compile time performance? If you make a repo we can make a list of common use cases and then design an API around that.

File: src/app/mod.rs (hunk 2 above)
Comment (ghost) [reply src/app/mod.rs:500]: I think in general the app structure is due for a cleanup, but I don't have a strong opinion at the moment on how. I'll work on this a bit later when I'm more free.

**Comment by antoyo** [comment]
It is ready to be merged?

**Comment by ghost** [comment]
I believe so

**Comment by antoyo** [comment]
The web extension doesn't compile: please fix that.

**Comment by ghost** [comment]
sorry about that, I built titanium but not the web extension and forgot, had to run out of the house

**Comment by antoyo** [comment]
The `match_pattern_all` function is not used (please check the warnings generated by the compiler).
Is this function intended for a future use?

**Comment by antoyo** [comment]
Thanks for your great work!
